### PR TITLE
MGDOBR-779: Make BridgeInstance error handling strategy editable

### PR DESCRIPTION
**Pull request by manstis**
[https://issues.redhat.com/browse/MGDOBR-779](https://issues.redhat.com/browse/MGDOBR-779)

Following the virtual Face to Face the approach taken in this PR was considered suitable.

Details of the Bridge instance status will be available when [MGDOBR-52](https://issues.redhat.com/browse/MGDOBR-52) is completed.

## PR owner checklist

Please make sure that your PR meets the following requirements:

- [x] Your code is properly formatted according to [this configuration](https://github.com/kiegroup/kogito-runtimes/tree/main/kogito-build/kogito-ide-config).  
  *Run `mvn clean verify -DskipTests` so that imports are correctly set.*
- [x] Your commit messages are clear and reference the JIRA issue e.g: "[MGDOBR-1] - $clear_explanation_of_what_you_did"
- [x] The layers in the `kustomize` folder have been updated accordingly.
- [x] All new functionality is tested
- [x] Pull Request title is properly formatted: `MGDOBR-XYZ Subject`
- [x] Pull Request contains link to the JIRA issue
- [x] Pull Request contains link to any dependent or related Pull Request
- [x] Pull Request contains description of the issue
- [x] Pull Request does not include fixes for issues other than the main ticket

## PR reviewer(s) checklist

- [ ] Check if code and Github action workflows have been modified and if the modification is safe
- [ ] If the modification is safe, add the `safe to test` label

<details>
<summary>
How to trigger pipelines and use the bots:
</summary>

- <b>Run the end to end pipeline</b>  
  Annotate the pull request with the label: `safe to test`. If you want to run the pipeline again, remove and add it again.

- <b>Rebase the pull request</b>  
  Comment with: `/rebase`.
- <b>Deploy BOT</b>
  Comment with `/deploy <dev,stable> when the PR has been merged to deploy to a target environment.

</details>

### CODE REVIEW

File: manager/src/main/java/com/redhat/service/smartevents/manager/ProcessorServiceImpl.java
Comment (manstis) [inline manager/src/main/java/com/redhat/service/smartevents/manager/ProcessorServiceImpl.java:169]: It is now possible to update an Error Handler Processor.

File: manager/src/main/java/com/redhat/service/smartevents/manager/BridgesServiceImpl.java
Comment (manstis) [inline manager/src/main/java/com/redhat/service/smartevents/manager/BridgesServiceImpl.java:142]: I have moved deletion of the Error Handler processor to `BridgeWorker` for consistency.

**Comment by r00ta** [comment]
Please do not merge this until I remove the label -> I'm checking that all the knative changes are working on main/dev

File: manager/src/main/java/com/redhat/service/smartevents/manager/BridgesServiceImpl.java
Comment (manstis) [inline manager/src/main/java/com/redhat/service/smartevents/manager/BridgesServiceImpl.java:109]: This is interesting....

To make things "simple" we cannot allow an Error Handler to be updated unless it is `READY`.

However Users know nothing of Error Handler processor provisioning as we chose to keep it internal.

Any better ideas to how to solve the problem?

We could make the `Bridge` stay in `PREPARING` until it's Error Handler is `READY` instead.

**Comment by manstis** [comment]
Tried and tested locally and it works as expected:
# Create a `Bridge` without an Error Handler - OK
# Update the `Bridge` and add an Error Handler - OK
# Update the `Bridge` and remove the Error Handler - OK

**Comment by manstis** [comment]
I need to write a few more Unit Tests for the latest commits.

File: manager/src/main/java/com/redhat/service/smartevents/manager/BridgesServiceImpl.java
Comment (sutaakar) [reply manager/src/main/java/com/redhat/service/smartevents/manager/BridgesServiceImpl.java:109]: Fine for me to keep `Bridge` in `PROVISIONING` state until all the components are fully provisioned.

File: manager/src/main/java/com/redhat/service/smartevents/manager/workers/resources/BridgeWorker.java
Comment (sutaakar) [inline manager/src/main/java/com/redhat/service/smartevents/manager/workers/resources/BridgeWorker.java:107]: a nitpick - maybe throw an exception in case there are several hidden Processors as it is not possible to identify which one is error handler?

File: manager/src/main/java/com/redhat/service/smartevents/manager/BridgesServiceImpl.java
Comment (danielezonca) [reply manager/src/main/java/com/redhat/service/smartevents/manager/BridgesServiceImpl.java:109]: We had an initial discussion about this with @kelvah because I agree that we need to make this time to apply a change visible to users.
I agree that we can leverage `Bridge` intermediate statuses to represent this.

File: shard-operator/src/main/java/com/redhat/service/smartevents/shard/operator/BridgeExecutorServiceImpl.java
Comment (sutaakar) [inline shard-operator/src/main/java/com/redhat/service/smartevents/shard/operator/BridgeExecutorServiceImpl.java:89]: In what case is this branch called? Why do we need this new code here?

File: manager/src/main/java/com/redhat/service/smartevents/manager/api/user/BridgesAPI.java
Comment (sutaakar) [inline manager/src/main/java/com/redhat/service/smartevents/manager/api/user/BridgesAPI.java:103]: Missing 404 response code

File: shard-operator/src/main/java/com/redhat/service/smartevents/shard/operator/BridgeExecutorServiceImpl.java
Comment (manstis) [reply shard-operator/src/main/java/com/redhat/service/smartevents/shard/operator/BridgeExecutorServiceImpl.java:89]: It _could_ happen if an attempt is made to re-provision a `Processor` without change.

I experienced this exact issue with `Bridge` on this PR (when updating an Error Handler the `Bridge` itself is not changed but our provisioning mechanism passes it through the Shard Operator by virtue of the `Bridge` moving from `ACCEPTED`->`PREPARING` and its dependencies being `READY`).

I'd err on the side of caution and keep this code in the `BridgeExecutorServiceImpl` too. We didn't know it was an issue for `Bridge`'s until this PR to update the Error Handler. It does not harm, but does prevent a potential issue in the future.

**Comment by sutaakar** [comment]
Code looks ok otherwise, briefly checked locally and seems to work properly. Will do more tests once comments are addressed.

File: manager/src/main/java/com/redhat/service/smartevents/manager/workers/resources/BridgeWorker.java
Comment (manstis) [reply manager/src/main/java/com/redhat/service/smartevents/manager/workers/resources/BridgeWorker.java:107]: I have refactored slightly to handle the first Error Handler processor found.

File: shard-operator/src/main/java/com/redhat/service/smartevents/shard/operator/BridgeExecutorServiceImpl.java
Comment (sutaakar) [reply shard-operator/src/main/java/com/redhat/service/smartevents/shard/operator/BridgeExecutorServiceImpl.java:89]: ok, fine for me

**Comment by manstis** [comment]
I have tested the following scenarios locally:
- Create a `Bridge` without Error Handler - OK
- Add an Error Handler to the `Bridge` - OK
- Update the Error Handler on the `Bridge` - OK
- Delete the Error Handler from the `Bridge` - OK
- Delete the `Bridge` - OK
- Create a `Bridge` with an Error Handler - OK
- Update the Error Handler on the `Bridge` - OK
- Delete the `Bridge` (and the Error Handler is deleted automatically) - OK

File: manager/src/main/java/com/redhat/service/smartevents/manager/BridgesServiceImpl.java
Comment (robpblake) [reply manager/src/main/java/com/redhat/service/smartevents/manager/BridgesServiceImpl.java]: Just to be clear on the implications for the Bridge `status` lifecycle here:

- If the error handler fails to deploy, is the containing Bridge instance marked as `FAILED`? If not, how does the user know if their error handler is broken if we failed to deploy it?
- If a user updates an error handler, and it fails to deploy, is the containing Bridge instance marked as `FAILED`?

In particular, that last scenario doesn't feel correct to me. I don't think we'd want to mark a bridge as `FAILED` if the user does a bad update to the error handler.

It feels like the error handler needs to be its own resource, perhaps at `bridges/{id}/error_handler` so that we can separate the lifecycle of the error_handler from the bridge instance on which is it deployed. That would mean that the user can determine if the error handler is working (or not) and update it without impacting the status of the Bridge e.g making a bad update to the error handler doesn't cause a bridge to go into a `FAILED` status.

Comment (manstis) [reply manager/src/main/java/com/redhat/service/smartevents/manager/BridgesServiceImpl.java]: Hi @robpblake well...... 

The original Error Handler proposal had the `Processors` listed a long with every other `SINK` or `SOURCE` `Processor` however "we" deemed it unsuitable for Users to be aware of internal implementation details and hence hid Error Handler `Processor`'s from view... 

That was OK when you could only create them.

Now we need to support updating the Error Handler definition (or even removing or adding it entirely) we have to prevent the operation if the underlying `Processor` is not `READY` otherwise we run into all sorts of race condition issues. 

This PR, whilst supporting updating the Error Handler, does change the life-cycle to exactly as you fear.

An Error Handler `Processor` is now "just" a dependency of the `Bridge`... hence, as with other dependencies (i.e. RHOC and `Processor`'s), if provisioning of the dependency fails the parent is moved into a `FAILED` state too. Meaning both of your horror scenarios are true.

It sounds like we need a meeting (you, @danielezonca, @kelvah, me and anyone else interest....)

File: manager/src/main/java/com/redhat/service/smartevents/manager/BridgesServiceImpl.java
Comment (sutaakar) [inline manager/src/main/java/com/redhat/service/smartevents/manager/BridgesServiceImpl.java:116]: `Objects.nonNull(updatedErrorHandler) ? updatedErrorHandler : null`
AFAIK that is the same as using `updatedErrorHandler` directly, right? Or is there some specific meaning for this construct?

File: manager/src/main/java/com/redhat/service/smartevents/manager/BridgesServiceImpl.java
Comment (sutaakar) [inline manager/src/main/java/com/redhat/service/smartevents/manager/BridgesServiceImpl.java:131]: AFAIK the work handling should be synchronized with https://github.com/5733d9e2be6485d52ffa08870cabdee0/sandbox/pull/920 (adjusting the PR once the other one is merged)

Comment (manstis) [reply manager/src/main/java/com/redhat/service/smartevents/manager/BridgesServiceImpl.java:131]: Correct. I would suggest #920 be merged first.

The _approach_ to editing Bridges is still not agreed across the team.

I _thought_ we was going to have agreement on Friday but it never came to be.

I'll move this back to draft.

File: manager/src/main/java/com/redhat/service/smartevents/manager/BridgesServiceImpl.java
Comment (manstis) [reply manager/src/main/java/com/redhat/service/smartevents/manager/BridgesServiceImpl.java:116]: hahaha, yes it is.. IDK what I was thinking. Changed.

**Comment by manstis** [comment]
I have repeated my earlier tests of the following scenarios locally:

- Create a Bridge without Error Handler - OK
- Add an Error Handler to the Bridge - OK
- Update the Error Handler on the Bridge - OK
- Delete the Error Handler from the Bridge - OK
- Delete the Bridge - OK
- Create a Bridge with an Error Handler - OK
- Update the Error Handler on the Bridge - OK
- Delete the Bridge (and the Error Handler is deleted automatically) - OK

**Comment by manstis** [comment]
@robpblake @danielezonca @sutaakar This PR is ready for review now.

**Review by sutaakar — APPROVED** [review]
Looks good to me

**Comment by manstis** [comment]
**NOTE TO SELF** When https://github.com/5733d9e2be6485d52ffa08870cabdee0/sandbox/pull/899 is merged this PR needs to be updated to prevent the Cloud Provider and Region being updated on a Bridge update...

File: manager/src/main/java/com/redhat/service/smartevents/manager/BridgesServiceImpl.java
Comment (jomarko) [inline manager/src/main/java/com/redhat/service/smartevents/manager/BridgesServiceImpl.java:119]: Could we please use `Objects.equals()` ? reading this class, not sure `existingDefinition` is `!= null`.

File: manager/src/main/java/com/redhat/service/smartevents/manager/ProcessorServiceImpl.java
Comment (jomarko) [inline manager/src/main/java/com/redhat/service/smartevents/manager/ProcessorServiceImpl.java:108]: wouldn't be more suitable to have some self explaining enum `GenerationStrategy long` rather than `long generation`?

File: manager/src/test/java/com/redhat/service/smartevents/manager/api/user/BridgesAPITest.java
Comment (jomarko) [inline manager/src/test/java/com/redhat/service/smartevents/manager/api/user/BridgesAPITest.java:352]: Is there also an error handler state that should cause bridge deletion to fail? if yes, should we cover it by test?

**Review by jomarko — COMMENTED** [review]
Thank you @manstis . See my comments inline and https://github.com/manstis/sandbox/pull/5/files

File: manager/src/main/java/com/redhat/service/smartevents/manager/BridgesServiceImpl.java
Comment (manstis) [reply manager/src/main/java/com/redhat/service/smartevents/manager/BridgesServiceImpl.java:119]: Done. I forgot your fondness of `Objects.equals(..)` (and I do not disagree, it is a good idea).

File: manager/src/main/java/com/redhat/service/smartevents/manager/ProcessorServiceImpl.java
Comment (manstis) [reply manager/src/main/java/com/redhat/service/smartevents/manager/ProcessorServiceImpl.java:108]: No, it needs to be a numerical type.

`generation` increases each time a `Bridge` is [saved](https://github.com/5733d9e2be6485d52ffa08870cabdee0/sandbox/pull/875/files#diff-66557df441afc3933a43c346564ae30380002fb47b8a0f12612ebcf2e71ece79R128). When updating a `Bridge` instance in` BridgeWorker` if the Error Handler `Processor` generation [is not equal to](https://github.com/5733d9e2be6485d52ffa08870cabdee0/sandbox/pull/875/files#diff-1319ec61fecd04bc61a3e44b2bffba77efe39a0885cf8d585aa9b01b97736859R122) the that of the `Bridge` it means the `Processor` is out of date and needs updating too. We then [set](https://github.com/5733d9e2be6485d52ffa08870cabdee0/sandbox/pull/875/files#diff-58ef592841ec60f16af57955f1413b4962e3111a85558c21e9f346bbf10ae4f9R122) the Error Handler `Processor` generation to be the same as that for the `Bridge.

Comment (jomarko) [reply manager/src/main/java/com/redhat/service/smartevents/manager/ProcessorServiceImpl.java:108]: Ok, then it is fair to no use enum. However I do not see any `+ 1` mechanism in `BridgeWorker` as you mentioned in `ConnectorWorker`

File: manager/src/test/java/com/redhat/service/smartevents/manager/api/user/BridgesAPITest.java
Comment (manstis) [reply manager/src/test/java/com/redhat/service/smartevents/manager/api/user/BridgesAPITest.java]: New test added.

**Review by jomarko — APPROVED** [review]

File: manager/src/main/java/com/redhat/service/smartevents/manager/ProcessorServiceImpl.java
Comment (manstis) [reply manager/src/main/java/com/redhat/service/smartevents/manager/ProcessorServiceImpl.java:108]: That's a great question!

I've tried to answer it at least three times; but I am failing! I recall I was setting the Error Handler generation to equal that of the `Bridge` to avoid unnecessary updates that I saw happening when testing; but I cannot remember the exact circumstances. It also leads me to then believe the same should be done in `ConnectorWorker` (set to that of the `Processor`) and not the increment as it does now. I am also concerned setting a `Processor` generation to that of the `Bridge` is wrong for regular processors not _tied_ to the life-cycle of the `Bridge`.

I need to investigate more..

Comment (manstis) [reply manager/src/main/java/com/redhat/service/smartevents/manager/ProcessorServiceImpl.java:108]: https://issues.redhat.com/browse/MGDOBR-975

**Comment by manstis** [comment]
@danielezonca @robpblake This is good to review/merge now.

I've added the checks to prevent a `Bridge` `CloudProvider` and/or `Region` from being updated.

File: manager/src/main/java/com/redhat/service/smartevents/manager/ProcessorService.java
Comment (robpblake) [inline manager/src/main/java/com/redhat/service/smartevents/manager/ProcessorService.java:23]: @manstis Does it simplify our life and make things clearer if we have a `Optional<Processor> getErrorHandler(String bridgeId)` method?  It seems a little confusing to me (I'm easily confused I appreciate) that in some places we're very explicit e.g create/update Error handler, but in other's we rely on knowledge of us understanding that the list of hidden processors exists and may contain the error handler.

It existed that way before this PR, so wonder if that's a follow up change we want to make if you think it sounds sensible.

cheers,

Rob

File: manager/src/main/java/com/redhat/service/smartevents/manager/workers/resources/BridgeWorker.java
Comment (robpblake) [inline manager/src/main/java/com/redhat/service/smartevents/manager/workers/resources/BridgeWorker.java:114]: @manstis We could extract this into a `getErrorHandler` method which returns the `Optional` as we have similar logic in the delete flow. Or see my previous comment about providing a `getErrorHandler` method on the `ProcessorService` which feels perhaps more appropriate.

File: manager/src/main/java/com/redhat/service/smartevents/manager/workers/resources/BridgeWorker.java
Comment (robpblake) [inline manager/src/main/java/com/redhat/service/smartevents/manager/workers/resources/BridgeWorker.java:84]: @manstis Checking my understanding here: The `Bridge` will enter the `FAILED` state only after exceeding the time limit for the provisioning process if the ErrorHandler fails to deploy?

If the `ErrorHandler` has failed to deploy, shouldn't we immediately mark the `Bridge` as failed within this loop as well?

cheers,

Rob

**Review by robpblake — COMMENTED** [review]
@manstis Couple of questions/comments from me. Thanks very much!

cheers,
Rob

File: manager/src/main/java/com/redhat/service/smartevents/manager/ProcessorServiceImpl.java
Comment (manstis) [reply manager/src/main/java/com/redhat/service/smartevents/manager/ProcessorServiceImpl.java:107]: Updated.

File: manager/src/main/java/com/redhat/service/smartevents/manager/ProcessorService.java
Comment (manstis) [reply manager/src/main/java/com/redhat/service/smartevents/manager/ProcessorService.java:23]: Done. It does avoid a duplication of code in `BridgeWorker`.

File: manager/src/main/java/com/redhat/service/smartevents/manager/workers/resources/BridgeWorker.java
Comment (manstis) [reply manager/src/main/java/com/redhat/service/smartevents/manager/workers/resources/BridgeWorker.java:114]: Done. It does avoid a duplication of code in `BridgeWorker`.

File: manager/src/main/java/com/redhat/service/smartevents/manager/workers/resources/BridgeWorker.java
Comment (manstis) [reply manager/src/main/java/com/redhat/service/smartevents/manager/workers/resources/BridgeWorker.java:84]: Well........ you have to go and open a whoop-ar5e can of worms; don't you :laughing: 

The code here would put a `Bridge`'s dependencies into `READY` so that the `Bridge` itself would continue to be re-provisioned by the Operator and the `Bridge` move into `READY` itself. This would happen if the Error Handler was either successfully updated or failed leading to a scenario where the `Bridge` is `READY` but Error Handler `FAILED`. 

Your comment suggests the preferred scenario is that, if the Error Handler fails we short-circuit and fail the `Bridge` too.

I've updated this PR to do the same. It is definitely consistent with parent-child dependencies elsewhere.

File: manager/src/main/java/com/redhat/service/smartevents/manager/ProcessorService.java
Comment (danielezonca) [inline manager/src/main/java/com/redhat/service/smartevents/manager/ProcessorService.java:33]: I like the `UserVisible`/`Hidden` naming to make clear that there are hidden processors, now this is not visible anymore but "implicit"

File: manager/src/main/java/com/redhat/service/smartevents/manager/workers/resources/BridgeWorker.java
Comment (robpblake) [reply manager/src/main/java/com/redhat/service/smartevents/manager/workers/resources/BridgeWorker.java]: @manstis That was my understanding of the outcome of the discussions with the other week at the F2F. I thought that we had agreed:

- Failure to deploy the Error Handler would result in the `Bridge` additionally entering the `FAILED` status, as the user has no direct visibility on the Error Handler
- The work you are doing to add a `status_message` to the `Bridge` resource in #968  would provide the user with a bit more information as to why the `Bridge` is in `FAILED` status.

Do you feel anything I'm suggesting here is not in alignment with those discussions?

cheers,

Rob

**Review by danielezonca — APPROVED** [review]

File: manager/src/main/java/com/redhat/service/smartevents/manager/ProcessorService.java
Comment (manstis) [reply manager/src/main/java/com/redhat/service/smartevents/manager/ProcessorService.java:33]: @danielezonca We now only have `ListResult<Processor> getProcessors(..)` (all `SINK` and `SOURCE`) and `Optional<Processor> getErrorHandler(..)`. Everywhere we were using `getHiddenProcessors(..)` we examining the list for the first `ERROR_HANDLER`. We only support defining one Error Handler at the moment so I think this change is more aligned.

File: manager/src/main/java/com/redhat/service/smartevents/manager/workers/resources/BridgeWorker.java
Comment (manstis) [reply manager/src/main/java/com/redhat/service/smartevents/manager/workers/resources/BridgeWorker.java]: No, no @robpblake I agree and have made the change in this PR.

**Review by robpblake — APPROVED** [review]

**Pull request merged**
